Compilation errors with Eigen 3.3.5

Our macOS builds started failing last week. I investigated a bit, and it seems that Eigen 3.3.5 was released last week (July 23, 2018). Building with Eigen 3.3.4 doesn't have this issue.

```
In file included from path/to/aikido/src/planner/parabolic/ParabolicTimer.cpp:1:
In file included from path/to/aikido/include/aikido/planner/parabolic/ParabolicTimer.hpp:4:
In file included from /usr/local/include/eigen3/Eigen/Dense:1:
In file included from /usr/local/include/eigen3/Eigen/Core:457:
/usr/local/include/eigen3/Eigen/src/Core/PlainObjectBase.h:927:5: error: static_assert failed
      "YOU_CALLED_A_DYNAMIC_SIZE_METHOD_ON_A_FIXED_SIZE_MATRIX_OR_VECTOR"
    EIGEN_STATIC_ASSERT_DYNAMIC_SIZE(Derived)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/eigen3/Eigen/src/Core/util/StaticAssert.h:150:3: note: expanded from macro
      'EIGEN_STATIC_ASSERT_DYNAMIC_SIZE'
  EIGEN_STATIC_ASSERT(TYPE::SizeAtCompileTime==Eigen::Dynamic, \
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/eigen3/Eigen/src/Core/util/StaticAssert.h:33:40: note: expanded from macro 'EIGEN_STATIC_ASSERT'
    #define EIGEN_STATIC_ASSERT(X,MSG) static_assert(X,#MSG);
                                       ^             ~
/usr/local/include/eigen3/Eigen/src/Core/PlainObjectBase.h:392:57: note: in instantiation of member function
      'Eigen::internal::conservative_resize_like_impl<Eigen::Matrix<double, 2, 1, 0, 2, 1>, Eigen::Matrix<double, 2,
      1, 0, 2, 1>, false>::run' requested here
      internal::conservative_resize_like_impl<Derived>::run(*this, rows, cols);
                                                        ^
/usr/local/include/eigen3/Eigen/src/SparseQR/SparseQR.h:638:11: note: in instantiation of member function
      'Eigen::PlainObjectBase<Eigen::Matrix<double, 2, 1, 0, 2, 1> >::conservativeResize' requested here
      res.conservativeResize(rows(), cols());
          ^
/usr/local/include/eigen3/Eigen/src/Core/ReturnByValue.h:62:42: note: in instantiation of function template
      specialization 'Eigen::SparseQR_QProduct<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>,
      Eigen::COLAMDOrdering<int> >, Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true>
      >::evalTo<Eigen::Matrix<double, 2, 1, 0, 2, 1> >' requested here
    { static_cast<const Derived*>(this)->evalTo(dst); }
                                         ^
/usr/local/include/eigen3/Eigen/src/Core/Assign.h:84:19: note: in instantiation of function template specialization
      'Eigen::ReturnByValue<Eigen::SparseQR_QProduct<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>,
      Eigen::COLAMDOrdering<int> >, Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true> >
      >::evalTo<Eigen::Matrix<double, 2, 1, 0, 2, 1> >' requested here
  other.derived().evalTo(derived());
                  ^
/usr/local/include/eigen3/Eigen/src/Core/PlainObjectBase.h:476:20: note: in instantiation of function template
      specialization 'Eigen::MatrixBase<Eigen::Matrix<double, 2, 1, 0, 2, 1>
      >::operator=<Eigen::SparseQR_QProduct<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>,
      Eigen::COLAMDOrdering<int> >, Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true> > >' requested
      here
      return Base::operator=(func);
                   ^
/usr/local/include/eigen3/Eigen/src/Core/Matrix.h:245:20: note: (skipping 4 contexts in backtrace; use
      -ftemplate-backtrace-limit=0 to see all)
      return Base::operator=(func);
                   ^
/usr/local/include/eigen3/Eigen/src/Core/PlainObjectBase.h:732:17: note: in instantiation of function template
      specialization 'Eigen::internal::call_assignment_no_alias<Eigen::Matrix<double, 2, 1, 0, 2, 1>,
      Eigen::Solve<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>, Eigen::COLAMDOrdering<int> >,
      Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true> >, Eigen::internal::assign_op<double, double>
      >' requested here
      internal::call_assignment_no_alias(this->derived(), other.derived(), internal::assign_op<Scalar,typenam...
                ^
/usr/local/include/eigen3/Eigen/src/Core/PlainObjectBase.h:537:7: note: in instantiation of function template
      specialization 'Eigen::PlainObjectBase<Eigen::Matrix<double, 2, 1, 0, 2, 1>
      >::_set_noalias<Eigen::Solve<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>, Eigen::COLAMDOrdering<int> >,
      Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true> > >' requested here
      _set_noalias(other);
      ^
/usr/local/include/eigen3/Eigen/src/Core/Matrix.h:379:9: note: in instantiation of function template specialization
      'Eigen::PlainObjectBase<Eigen::Matrix<double, 2, 1, 0, 2, 1>
      >::PlainObjectBase<Eigen::Solve<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>, Eigen::COLAMDOrdering<int>
      >, Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true> > >' requested here
      : Base(other.derived())
        ^
path/to/aikido/include/aikido/common/detail/Spline-impl.hpp:407:11: note: in instantiation of
      function template specialization 'Eigen::Matrix<double, 2, 1, 0, 2,
      1>::Matrix<Eigen::Solve<Eigen::SparseQR<Eigen::SparseMatrix<double, 0, int>, Eigen::COLAMDOrdering<int> >,
      Eigen::Block<Eigen::Matrix<double, 2, -1, 0, 2, -1>, 2, 1, true> > >' requested here
        = solver.solve(mB.col(ioutput));
          ^
path/to/aikido/src/planner/parabolic/ParabolicTimer.cpp:66:33: note: in instantiation of member
      function 'aikido::common::SplineProblem<double, int, 2, -1, 2>::fit' requested here
    const auto spline = problem.fit();
```

This doesn't seem to be affecting Xenial builds, which makes sense since the Xenial version is pinned to 3.3~beta1-2.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -60,6 +60,7 @@
 
   * Fixed Eigen memory alignment issues on 32bit Ubuntu: [#368](https://github.com/personalrobotics/aikido/pull/368)
   * Defined optional dependencies: [#376](https://github.com/personalrobotics/aikido/pull/376)
+  * Fixed compilation bug with Eigen 3.3.5: [#452](https://github.com/personalrobotics/aikido/pull/452)
 
 ### 0.2.0 (2018-01-09)
 

--- a/include/aikido/common/detail/Spline-impl.hpp
+++ b/include/aikido/common/detail/Spline-impl.hpp
@@ -403,7 +403,11 @@ auto SplineProblem<Scalar, Index, _NumCoefficients, _NumOutputs, _NumKnots>::
   for (Index ioutput = 0; ioutput < mNumOutputs; ++ioutput)
   {
     // Solve for the spline coefficients for each output dimension.
-    Eigen::Matrix<Scalar, DimensionAtCompileTime, 1> solutionVector
+    //
+    // TODO: As of Eigen 3.3.5, the output type of SparseQR::solve() is not
+    // assignable to a fixed size vector, so we use Eigen::Dynamic here instead
+    // of DimensionAtCompileTime.
+    Eigen::Matrix<Scalar, Eigen::Dynamic, 1> solutionVector
         = solver.solve(mB.col(ioutput));
 
     // Split the coefficients by segment.

